source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst

```diff
@@ -2072,14 +2118,36 @@ then it is probably down.
 Use error messages to detect "not master" and "node is recovering"
 ''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
 
-An alternative idea is to determine all relevant server error codes,
-instead of searching for substrings in the error message.
-But for "not master" and "node is recovering" errors,
-driver authors have found the substrings to be **more** stable
-than error codes.
+When error codes are not available, error messages are checked for the
```

Comment:
By "error codes are not available" do you mean ONLY the scenario where the "code" field is missing?

I'm guessing you might actually mean to include the scenario where the "code" field is present but the error code is not one of the ones documented in these changes, and that your intention is that in that case as a fallback we ALSO inspect the message string for certain substrings.

Reply:
Actually I did mean only the scenario when the code field is present. I was under the impression that there'd be no point to checking the message if there is an error code present.

Reply:
The text you replaced says that drivers found the substrings more reliable than the error codes (that's NOT saying that the code was missing, it's saying that code changed from version to version of the server). If that's the case we SHOULD check the substring even when the error code is present (assuming we don't recognize the code).

Is it even possible for the server to return { ok : 0 } without a code field?

Reply:
Sure, until recently it was epidemic for the server to return { ok: 0 } without a code field. An insert on a secondary running 3.0:
```
db.runCommand({insert:'collection', documents: [{}]})
{ "note" : "from execCommand", "ok" : 0, "errmsg" : "not master" }
```
No code. For a half-dozen more examples:

https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring-tests.rst

Reply:
I propose that if an error has a code not specifically listed in this spec, the client MUST parse the error string for "not master" or "node is recovering" according to the rules here, no matter whether the error has a code or not. That's consistent with the rules we've used so far.

Reply:
> Is it even possible for the server to return { ok : 0 } without a code field?

Yep, see the comment above: https://github.com/mongodb/specifications/pull/299#discussion_r178420476

Maybe drivers should just check these error codes *in addition* to the old error message string logic.

Reply:
Sounds good to me, I updated the pseudo code and description to first check error code. If the error code is omitted or not in the list, fall back to the substring check.